// symstr/Group.js
var Mod = require("mod42/v1a");
var View = require("view42/v1");

var Item = Mod.extend({
	name: "Item",
	value: "new item",
	View: View.extend({
		name: "ItemView",
		initialize: function(){
			this.render();
			// this.update();
		},
		render: function(){
			this.append({
				value: View(this.item.value),
				btns: View().append({
					dupe: View("d").click(function(){
						this.parent.parent.item.dupe();
					}),
					global: View("g").click(function(){
						window.view = this.parent.parent;
					}),
					editBtn: View("e").click(function(){
						this.parent.parent.edit();
					}) 
				})
			});
		},
		edit: function(){
			this.value.el.setAttribute("contenteditable", true);
			this.value.el.focus();

			if (!this.saveHandler){
				this.saveHandler = function(e){
					console.log("press");
					if (e.key === "Enter"){
						e.preventDefault();
						this.el.removeEventListener("keypress", this.saveHandler);
						this.saveHandler = false;
						this.save();
						this.stopEditing();
					}
				}.bind(this);

				this.el.addEventListener("keypress", this.saveHandler);
			}
		},
		save: function(){
			this.item.set_value(this.value.el.innerHTML);
		},
		stopEditing: function(){
			// this.
			this.value.el.removeAttribute("contenteditable");
		},
		update: function(){
			this.value.el.innerHTML = this.item.value;
		},
		
	}),
	initialize: function(){
		this.views = [];
		this.view = this.render(); // main view?
	},
	render: function(){
		var view = new this.View({
			item: this
		});
		this.views.push(view);
		return view;
	},
	set_value: function(newValue){
		this.value = newValue;

		for (var i = 0; i < this.views.length; i++){
			this.views[i].update();
		}
	},
	dupe: function(){
		var view = this.render();
		this.group.view.el.appendChild(view.el);
	}
});

var GroupView = View.extend({
	name: "GroupView",
	render: function(){
		this.append({
			btn: View("add")
		});

		this.btn.click(function(){
			this.parent.group.new();
		});
	},
	// newBlankItem: function(){
	// 	this
	// 	this.append({
	// 		newItem: View("new item")
	// 	});

	// 	this.newItem.focus();
	// }
	// update: function(){
	// 	this.empty();
	// 	for (var i = 0; i < this.items.length; i++){
	// 		this.
	// 	}
	// }
});

var Group = module.exports = Mod.extend({
	name: "Group",
	View: GroupView,
	initialize: function(){
		this.items = [];
		this.render();
	},
	render: function(){
		this.view = new this.View({
			group: this
		});
	},
	new: function(){
		var item = new Item({
			group: this
		});
		this.items.push(item);
		this.view.el.appendChild(item.view.el);
		item.view.edit();
	}
});
